File: Problems/435.non-overlapping-intervals.py
# title: Non-overlapping Intervals
# timestamp: 2025-01-10 07:15:25
# problemUrl: https://leetcode.com/problems/non-overlapping-intervals/
# memory: 51.5 MB
# runtime: 67 ms

class Solution:
    def eraseOverlapIntervals(self, intervals: List[List[int]]) -> int:
        intervals.sort(key=lambda i: i[1])
        end = -10**9
        count = 0
        for interval in intervals:
            if end <= interval[0]:
                end = interval[1]
                continue
            count += 1
            
        return count

File: Problems/435.non-overlapping-intervals.go
// title: Non-overlapping Intervals
// timestamp: 2024-04-23 16:48:22
// problemUrl: https://leetcode.com/problems/non-overlapping-intervals/
// memory: 16.8 MB
// runtime: 179 ms

import (
    "sort"
)

func eraseOverlapIntervals(intervals [][]int) int {
    sort.Slice(intervals, func(i, j int) bool {
        return intervals[i][0] < intervals[j][0]
    })

    const removed = -100000
    var count int
    for i, j := 0, 1; j < len(intervals); {
        if intervals[i][0] == removed {
            i++
            continue
        }
        if intervals[j][0] == removed {
            j++
            continue
        }
        if intervals[j][0] >= intervals[i][1] {
            i++
            j++
            continue
        }
        //remove one
        count++
        if intervals[i][1] < intervals[j][1] {
            intervals[j][0] = removed
            j++
            continue
        }
        intervals[i][0] = removed
        i++
        j++
    }

    return count
}